dbr quant: enable reference module support for torch.qint32

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#73493 dbr quant: enable reference module support for torch.qint32**
* #73492 dbr quant: insert activation obs explicitly, instead of relying on hooks

Summary:

This PR enables basic support for reference modules in DBR quant.
For now, the support is limited to:
1. modules that have reference versions defined only (no functions)
2. torch.qint32 dtype only

Currently, the reference module logic is enabled whenever dtype is
torch.qint32. This is done because this is needed the earliest for
the first use case. A future PR will support more dtypes and also
add the `is_reference` flag to the API.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_conv_int32_reference_model
```

Differential Revision: [D34520759](https://our.internmc.facebook.com/intern/diff/D34520759)

vkuzo committed:
dbr quant: enable reference module support for torch.qint32

Summary:

This PR enables basic support for reference modules in DBR quant.
For now, the support is limited to:
1. modules that have reference versions defined only (no functions)
2. torch.qint32 dtype only

Currently, the reference module logic is enabled whenever dtype is
torch.qint32. This is done because this is needed the earliest for
the first use case. A future PR will support more dtypes and also
add the `is_reference` flag to the API.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_conv_int32_reference_model
```

TODO finish this

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34520759).

vkuzo committed:
Update on "dbr quant: enable reference module support for torch.qint32"


Summary:

This PR enables basic support for reference modules in DBR quant.
For now, the support is limited to:
1. modules that have reference versions defined only (no functions)
2. torch.qint32 dtype only

Currently, the reference module logic is enabled whenever dtype is
torch.qint32. This is done because this is needed the earliest for
the first use case. A future PR will support more dtypes and also
add the `is_reference` flag to the API.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_conv_int32_reference_model
```

Differential Revision: [D34520759](https://our.internmc.facebook.com/intern/diff/D34520759)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/_dbr/module_swap_utils.py:
```diff
@@ -58,6 +63,18 @@ def _swap_child_modules(
+            # the is_reference API flag instead of doing this automatically.
+            for orig_mod_type, _ref_mod_type in \
+                    DEFAULT_REFERENCE_STATIC_QUANT_MODULE_MAPPINGS.items():
+                if not isinstance(mod, orig_mod_type):  # type: ignore[arg-type]
```
can we do this check in `swap_module`? not sure why we need to have a for loop here

vkuzo replied:
I see, looks like `swap_module` doesn't actually swap if the mapping is missing, maybe we can consider renaming it in the future.

jerryzh168 replied:
yeah renaming sounds good, it would make it clearer, like "maybe_swap_module", we need to redesign this api anyways to support swapping for multiple quantization modes anyways in the future I think, we can do it later

jerryzh168 reviewed (approved):
looks good!

vkuzo committed:
Update on "dbr quant: enable reference module support for torch.qint32"


Summary:

This PR enables basic support for reference modules in DBR quant.
For now, the support is limited to:
1. modules that have reference versions defined only (no functions)
2. torch.qint32 dtype only

Currently, the reference module logic is enabled whenever dtype is
torch.qint32. This is done because this is needed the earliest for
the first use case. A future PR will support more dtypes and also
add the `is_reference` flag to the API.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_conv_int32_reference_model
```

Differential Revision: [D34520759](https://our.internmc.facebook.com/intern/diff/D34520759)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34520759).

vkuzo committed:
Update on "dbr quant: enable reference module support for torch.qint32"


Summary:

This PR enables basic support for reference modules in DBR quant.
For now, the support is limited to:
1. modules that have reference versions defined only (no functions)
2. torch.qint32 dtype only

Currently, the reference module logic is enabled whenever dtype is
torch.qint32. This is done because this is needed the earliest for
the first use case. A future PR will support more dtypes and also
add the `is_reference` flag to the API.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_conv_int32_reference_model
```

Differential Revision: [D34520759](https://our.internmc.facebook.com/intern/diff/D34520759)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34520759).

vkuzo committed:
Update on "dbr quant: enable reference module support for torch.qint32"


Summary:

This PR enables basic support for reference modules in DBR quant.
For now, the support is limited to:
1. modules that have reference versions defined only (no functions)
2. torch.qint32 dtype only

Currently, the reference module logic is enabled whenever dtype is
torch.qint32. This is done because this is needed the earliest for
the first use case. A future PR will support more dtypes and also
add the `is_reference` flag to the API.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_conv_int32_reference_model
```

Differential Revision: [D34520759](https://our.internmc.facebook.com/intern/diff/D34520759)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34520759).
